lightning/src/ln/onion_utils.rs

```diff
@@ -1777,8 +1820,19 @@ impl_writeable_tlv_based_enum!(HTLCFailReasonRepr,
 		(_unused, err, (static_value, msgs::OnionErrorPacket { data: data.ok_or(DecodeError::InvalidValue)?, attribution_data })),
 	},
 	(1, Reason) => {
-		(0, failure_code, required),
+		(0, _failure_code, (legacy, u16,
+			|r: &HTLCFailReasonRepr| Some(r.clone()) )),
 		(2, data, required_vec),
+		// failure_code was required, and is replaced by reason so any time we do not have a
+		// reason available failure_code will be Some so we can require reason.
+		(4, reason, (default_value,
+			if let Some(code) = _failure_code {
+				let failure_reason: LocalHTLCFailureReason = code.into();
+				RequiredWrapper::from(failure_reason)
+			} else {
+				reason
+			}
```

Comment:
I think for safety we should return the unknown failure reason here even though (as you note above) the failure_code should always be set in this situation

Reply:
Changed this to return `InvalidValue` if `reason` is `None` and we don't have a `_failure_code`, similar to the way it's been done for `attribution_data` above.

Open to other approaches here too! Still getting comfortable w/ these macros.

Reply:
Makes sense, thanks! Definitely a bit of a learning curve to these macros at times